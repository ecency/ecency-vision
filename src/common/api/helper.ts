import defaults from "../constants/defaults.json";

export const apiBase = (endpoint: string): string => `${defaults.base}${endpoint}`;

export const apiBaseImage = (endpoint: string): string => `${defaults.imageServer}${endpoint}`;
