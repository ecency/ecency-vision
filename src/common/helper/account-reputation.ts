const isHumanReadable = (input: number): boolean => {
  return Math.abs(input) > 0 && Math.abs(input) <= 100;
};

export default (input: string | number): number => {
  if (typeof input === 'number' && isHumanReadable(input)) {
    return Math.floor(input);
  }

  if (typeof input === 'string') {
    input = Number(input);

    if (isHumanReadable(input)) {
      return Math.floor(input);
    }
  }

  if (input === 0) {
    return 25;
  }

  let neg = false;

  if (input < 0) neg = true;

  let reputationLevel = Math.log10(Math.abs(input));
  reputationLevel = Math.max(reputationLevel - 9, 0);

  if (reputationLevel < 0) reputationLevel = 0;

  if (neg) reputationLevel *= -1;

  reputationLevel = reputationLevel * 9 + 25;

  return Math.floor(reputationLevel);
};
