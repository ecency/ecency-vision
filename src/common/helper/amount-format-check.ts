export default (v: string) => /^\d{1,3}(,?\d{3})*(\.(\d{3},?)*\d{1,3})?$/.test(v);
