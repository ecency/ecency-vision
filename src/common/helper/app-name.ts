import stripTags from "../util/strip-tags";

interface ObjInput {
  name?: string;
}

export default (input: string | null | undefined | ObjInput) => {
  if (!input) {
    return "";
  }

  if (typeof input === "string") {
    return stripTags(input);
  }

  if (typeof input === "object" && input.name !== undefined) {
    return stripTags(input.name);
  }

  return "";
};
