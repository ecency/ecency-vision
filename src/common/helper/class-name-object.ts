export function classNameObject(obj: Record<string, boolean | undefined>): string {
  return Object.entries(obj)
    .filter(([className, isActive]) => !!isActive)
    .map(([className]) => className)
    .join(" ");
}
