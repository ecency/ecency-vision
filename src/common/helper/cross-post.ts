import {postBodySummary} from '@ecency/render-helper';
import {Entry} from '../store/entries/types';

export const crossPostMessage = (body: string) => {
  const crossPostRegex =
    /^This is a cross post of \[@(.*?)\/(.*?)\]\(\/.*?@.*?\/.*?\) by @.*?\.<br>/;

  if (body.match(crossPostRegex)) {
    const message = body.replace(crossPostRegex, '');
    return postBodySummary(message);
  }

  return null;
};

export const makeCrossPostMessage = (
  entry: Entry,
  poster: string,
  message: string,
) => {
  return `This is a cross post of [@${entry.author}/${entry.permlink}](/${entry.category}/@${entry.author}/${entry.permlink}) by @${poster}.<br><br>${message}`;
};
