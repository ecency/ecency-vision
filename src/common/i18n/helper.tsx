import React from "react";

import xss from "xss";

import { _t } from "./index";

export const HTML_WHITELIST = {
  a: ["class", "href", "target", "rel"],
  span: ["class"],
  h3: [],
  p: [],
  strong: [],
  b: [],
  img: ["src", "alt"],
  ul: [],
  li: [],
  br: []
};

export const safeHtml = (html: string): string => {
  return xss(html, {
    whiteList: HTML_WHITELIST,
    stripIgnoreTagBody: true,
    stripIgnoreTag: true,
    css: false
  });
};

interface Props {
  k: string;
  args?: {};
  children: JSX.Element;
}

export class Tsx extends React.Component<Props> {
  render() {
    const { k, args } = this.props;

    const { children } = this.props;

    const html = safeHtml(_t(k, args));

    return React.cloneElement(children, { dangerouslySetInnerHTML: { __html: html } });
  }
}
