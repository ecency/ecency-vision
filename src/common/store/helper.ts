import {Store} from 'redux';

import i18n from 'i18next';

import {AppState} from './index';
import {ActiveUser, UserPoints} from './active-user/types';
import {
  loginAct as loginActiveUser,
  logoutAct as logoutActiveUser,
  updateAct as updateActiveUserAct,
} from './active-user';

import {getAccount, getDynamicProps} from '../api/hive';
import {getPoints, usrActivity, getPromotedEntries} from '../api/private-api';
import {reloadAct as reloadUsers} from './users';
import {fetchedAct as loadDynamicProps} from './dynamic-props';
import {fetchedAct as entriesFetchedAct} from './entries';
import {
  setCurrencyAct as setCurrency,
  muteNotificationsAct as muteNotifications,
  setLangAct as setLang,
  setNsfwAct as setNsfw,
} from './global';

import {getCurrencyRate} from '../api/misc';

import currencies from '../constants/currencies.json';

import * as ls from '../../common/util/local-storage';

import currencySymbol from '../../common/helper/currency-symbol';

import {AppWindow} from '../../client/window';

declare var window: AppWindow;

export const activeUserMaker = (
  name: string,
  points: string = '0.000',
  uPoints: string = '0.000',
): ActiveUser => {
  return {
    username: name,
    data: {name},
    points: {
      points,
      uPoints,
    },
  };
};

export const activeUserUpdater = async (store: Store<AppState>) => {
  const state = store.getState();
  if (state.activeUser) {
    const {username} = state.activeUser;

    let account;
    try {
      account = await getAccount(username);
    } catch (e) {
      return;
    }

    let points: UserPoints;
    try {
      const p = await getPoints(username);
      points = {points: p.points, uPoints: p.unclaimed_points};
    } catch (e) {
      points = {
        points: '0.000',
        uPoints: '0.000',
      };
    }

    store.dispatch(updateActiveUserAct(account, points));
  }
};

export const syncActiveUser = (store: Store<AppState>) => {
  const state = store.getState();

  const activeUser = ls.get('active_user');

  // logout
  if (!activeUser && state.activeUser) {
    store.dispatch(logoutActiveUser());
    return;
  }

  // login
  if (activeUser && !state.activeUser) {
    store.dispatch(loginActiveUser());
    activeUserUpdater(store).then();
    return;
  }

  // switch
  if (
    activeUser &&
    state.activeUser &&
    activeUser !== state.activeUser.username
  ) {
    store.dispatch(loginActiveUser());
    activeUserUpdater(store).then();
    return;
  }
};

export const clientStoreTasks = (store: Store<AppState>) => {
  // To use in places where we can't access to store
  window.usePrivate = store.getState().global.usePrivate;

  // Initial state from browser's local storage
  store.dispatch(reloadUsers());
  store.dispatch(loginActiveUser());

  // Load dynamic props
  getDynamicProps().then(resp => {
    store.dispatch(loadDynamicProps(resp));
  });

  // Update active user in interval
  activeUserUpdater(store).then();
  setInterval(() => {
    activeUserUpdater(store).then();
  }, 60 * 1000);

  // Active user sync between tabs / windows
  setInterval(() => {
    syncActiveUser(store);
  }, 2000);

  // Do check-in in interval
  const checkIn = () => {
    const state = store.getState();
    if (state.activeUser) {
      usrActivity(state.activeUser?.username!, 10).then();
    }
  };

  // wait for initial user sync
  setTimeout(() => {
    checkIn();
  }, 5000);
  setInterval(checkIn, 1000 * 60 * 15 + 8);

  // Inject / update promoted entries to store
  const promotedEntries = () => {
    getPromotedEntries().then(r => {
      store.dispatch(entriesFetchedAct('__promoted__', r, false));
    });
  };
  promotedEntries();
  setInterval(promotedEntries, 1000 * 60 * 5);

  // Currency
  const currency = ls.get('currency');
  if (currency && currencies.find(x => x.id === currency)) {
    const symbol = currencySymbol(currency);
    getCurrencyRate(currency).then(rate => {
      store.dispatch(setCurrency(currency, rate, symbol));
    });
  }

  // Notifications
  if (ls.get('notifications') === false) {
    store.dispatch(muteNotifications());
  }

  // Language
  const lang = ls.get('lang');
  if (lang) {
    if (store.getState().global.lang !== lang) {
      i18n.changeLanguage(lang).then(() => {
        store.dispatch(setLang(lang));
      });
    }
  }

  // NSFW
  if (ls.get('nsfw') === true) {
    store.dispatch(setNsfw(true));
  }
};
