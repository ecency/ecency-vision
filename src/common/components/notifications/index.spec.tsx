import React from 'react';

import TestRenderer from 'react-test-renderer';
import {createBrowserHistory} from 'history';

import {globalInstance} from "../../helper/test-helper";

import {NotificationListItem} from './index';

import {
    apiVoteNotification,
    apiMentionNotification,
    apiFollowNotification,
    apiUnfollowNotification,
    apiReplyNotification,
    apiReblogNotification,
    apiTransferNotification
} from "../../helper/test-helper";


describe('(1) NotificationListItem', () => {

    const defProps = {
        global: globalInstance,
        history: createBrowserHistory(),
        markNotifications: () => {

        },
        addAccount: () => {

        },
        toggleUIProp: () => {

        }
    }

    it('(1) Vote ', () => {
        const props = {...defProps, notification: apiVoteNotification,};
        const renderer = TestRenderer.create(
            <NotificationListItem {...props}/>
        );

        expect(renderer.toJSON()).toMatchSnapshot();
    });

    it('(2) Mention ', () => {
        const props = {...defProps, notification: apiMentionNotification,};
        const renderer = TestRenderer.create(
            <NotificationListItem {...props}/>
        );

        expect(renderer.toJSON()).toMatchSnapshot();
    });

    it('(3) Follow ', () => {
        const props = {...defProps, notification: apiFollowNotification,};
        const renderer = TestRenderer.create(
            <NotificationListItem {...props}/>
        );

        expect(renderer.toJSON()).toMatchSnapshot();
    });

    it('(4) Unfollow ', () => {
        const props = {...defProps, notification: apiUnfollowNotification,};
        const renderer = TestRenderer.create(
            <NotificationListItem {...props}/>
        );

        expect(renderer.toJSON()).toMatchSnapshot();
    });

    it('(5) Reply ', () => {
        const props = {...defProps, notification: apiReplyNotification,};
        const renderer = TestRenderer.create(
            <NotificationListItem {...props}/>
        );

        expect(renderer.toJSON()).toMatchSnapshot();
    });

    it('(6) Reblog ', () => {
        const props = {...defProps, notification: apiReblogNotification,};
        const renderer = TestRenderer.create(
            <NotificationListItem {...props}/>
        );

        expect(renderer.toJSON()).toMatchSnapshot();
    });

    it('(7) Transfer ', () => {
        const props = {...defProps, notification: apiTransferNotification,};
        const renderer = TestRenderer.create(
            <NotificationListItem {...props}/>
        );

        expect(renderer.toJSON()).toMatchSnapshot();
    });
})

