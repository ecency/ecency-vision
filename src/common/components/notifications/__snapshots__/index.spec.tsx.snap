// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(1) NotificationListItem (1) Vote  1`] = `
<div
  className="list-item not-read"
  title="2020-07-24T06:50:48+00:00"
>
  <div
    className="item-inner "
  >
    <div
      className="item-control "
    >
      <span
        className="mark-read"
        onClick={[Function]}
        title="Mark as Read"
      />
    </div>
    <div
      className="source"
    >
      <a
        className="source-avatar"
        href="/@fooo10"
        onClick={[Function]}
      >
        <span
          className="user-avatar medium"
          style={
            Object {
              "backgroundImage": "url(https://images.ecency.com/u/fooo10/avatar/medium)",
            }
          }
        />
      </a>
    </div>
    <div
      className="item-content"
    >
      <div
        className="first-line"
      >
        <a
          className="source-name"
          href="/@fooo10"
          onClick={[Function]}
        >
           
          fooo10
        </a>
        <span
          className="item-action"
        >
          100% voted your post:
        </span>
      </div>
      <div
        className="second-line"
      >
        <a
          className="post-link"
          href="/category/@good-karma/update-on-hivesigner-integration-into-condenser"
          onClick={[Function]}
        >
          update-on-hivesigner-integration-into-condenser
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`(1) NotificationListItem (2) Mention  1`] = `
<div
  className="list-item not-read"
  title="2020-07-24T01:10:09+00:00"
>
  <div
    className="item-inner "
  >
    <div
      className="item-control "
    >
      <span
        className="mark-read"
        onClick={[Function]}
        title="Mark as Read"
      />
    </div>
    <div
      className="source"
    >
      <a
        className="source-avatar"
        href="/@ecency.stats"
        onClick={[Function]}
      >
        <span
          className="user-avatar medium"
          style={
            Object {
              "backgroundImage": "url(https://images.ecency.com/u/ecency.stats/avatar/medium)",
            }
          }
        />
      </a>
    </div>
    <div
      className="item-content"
    >
      <div
        className="first-line"
      >
        <a
          className="source-name"
          href="/@ecency.stats"
          onClick={[Function]}
        >
           
          ecency.stats
        </a>
        <span
          className="item-action"
        >
          mentioned you in:
        </span>
      </div>
      <div
        className="second-line"
      >
        <a
          className="post-link"
          href="/category/@ecency.stats/daily-activities-points-2020-07-24"
          onClick={[Function]}
        >
          daily-activities-points-2020-07-24
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`(1) NotificationListItem (3) Follow  1`] = `
<div
  className="list-item not-read"
  title="2020-07-24T06:08:24+00:00"
>
  <div
    className="item-inner "
  >
    <div
      className="item-control "
    >
      <span
        className="mark-read"
        onClick={[Function]}
        title="Mark as Read"
      />
    </div>
    <div
      className="source"
    >
      <a
        className="source-avatar"
        href="/@fooo10"
        onClick={[Function]}
      >
        <span
          className="user-avatar medium"
          style={
            Object {
              "backgroundImage": "url(https://images.ecency.com/u/fooo10/avatar/medium)",
            }
          }
        />
      </a>
    </div>
    <div
      className="item-content"
    >
      <div
        className="first-line"
      >
        <a
          className="source-name"
          href="/@fooo10"
          onClick={[Function]}
        >
           
          fooo10
        </a>
      </div>
      <div
        className="second-line"
      >
        <span
          className="follow-label"
        >
          followed
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`(1) NotificationListItem (4) Unfollow  1`] = `
<div
  className="list-item"
  title="2020-07-19T05:07:09+00:00"
>
  <div
    className="item-inner "
  >
    <div
      className="item-control "
    />
    <div
      className="source"
    >
      <a
        className="source-avatar"
        href="/@fooo102"
        onClick={[Function]}
      >
        <span
          className="user-avatar medium"
          style={
            Object {
              "backgroundImage": "url(https://images.ecency.com/u/fooo102/avatar/medium)",
            }
          }
        />
      </a>
    </div>
    <div
      className="item-content"
    >
      <div
        className="first-line"
      >
        <a
          className="source-name"
          href="/@fooo102"
          onClick={[Function]}
        >
           
          fooo102
        </a>
      </div>
      <div
        className="second-line"
      >
        <span
          className="unfollow-label"
        >
          unfollowed
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`(1) NotificationListItem (5) Reply  1`] = `
<div
  className="list-item"
  title="2020-07-21T16:02:00+00:00"
>
  <div
    className="item-inner "
  >
    <div
      className="item-control "
    />
    <div
      className="source"
    >
      <a
        className="source-avatar"
        href="/@dunsky"
        onClick={[Function]}
      >
        <span
          className="user-avatar medium"
          style={
            Object {
              "backgroundImage": "url(https://images.ecency.com/u/dunsky/avatar/medium)",
            }
          }
        />
      </a>
    </div>
    <div
      className="item-content"
    >
      <div
        className="first-line"
      >
        <a
          className="source-name"
          href="/@dunsky"
          onClick={[Function]}
        >
           
          dunsky
        </a>
        <span
          className="item-action"
        >
          replied to your post:
        </span>
        <div
          className="vert-separator"
        />
        <a
          className="post-link"
          href="/category/@good-karma/re-dunsky-2020713t164324912z"
          onClick={[Function]}
        >
          re-dunsky-2020713t164324912z
        </a>
      </div>
      <div
        className="second-line"
      >
        <a
          className="markdown-view mini-markdown reply-body"
          href="/category/@dunsky/re-good-karma-2020721t1915968z"
          onClick={[Function]}
        >
          Testing replies from Ecency web interface ðŸ˜Ž
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`(1) NotificationListItem (6) Reblog  1`] = `
<div
  className="list-item not-read"
  title="2020-07-23T13:36:06+00:00"
>
  <div
    className="item-inner "
  >
    <div
      className="item-control "
    >
      <span
        className="mark-read"
        onClick={[Function]}
        title="Mark as Read"
      />
    </div>
    <div
      className="source"
    >
      <a
        className="source-avatar"
        href="/@foo220"
        onClick={[Function]}
      >
        <span
          className="user-avatar medium"
          style={
            Object {
              "backgroundImage": "url(https://images.ecency.com/u/foo220/avatar/medium)",
            }
          }
        />
      </a>
    </div>
    <div
      className="item-content"
    >
      <div
        className="first-line"
      >
        <a
          className="source-name"
          href="/@foo220"
          onClick={[Function]}
        >
           
          foo220
        </a>
        <span
          className="item-action"
        >
          reblogged your post:
        </span>
      </div>
      <div
        className="second-line"
      >
        <a
          className="post-link"
          href="/category/@good-karma/update-on-hivesigner-integration-into-condenser"
          onClick={[Function]}
        >
          update-on-hivesigner-integration-into-condenser
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`(1) NotificationListItem (7) Transfer  1`] = `
<div
  className="list-item"
  title="2020-07-20T09:30:21+00:00"
>
  <div
    className="item-inner "
  >
    <div
      className="item-control "
    />
    <div
      className="source"
    >
      <a
        className="source-avatar"
        href="/@bar212"
        onClick={[Function]}
      >
        <span
          className="user-avatar medium"
          style={
            Object {
              "backgroundImage": "url(https://images.ecency.com/u/bar212/avatar/medium)",
            }
          }
        />
      </a>
    </div>
    <div
      className="item-content"
    >
      <div
        className="first-line"
      >
        <a
          className="source-name"
          href="/@bar212"
          onClick={[Function]}
        >
           
          bar212
        </a>
        <span
          className="item-action"
        >
          sent you
           
          <span
            className="transfer-amount"
          >
            900.000 HBD
          </span>
        </span>
      </div>
      
    </div>
  </div>
</div>
`;
